src/src/pam_oddjob_mkhomedir.8.in
.TH pam_oddjob_mkhomedir 8 "29 May 2008" "oddjob Manual"

.SH NAME
pam_oddjob_mkhomedir.so - PAM module to create users home directory via oddjob

.SH SYNOPSIS
session optional /lib/security/pam_oddjob_mkhomedir.so

.SH DESCRIPTION
The \fIpam_oddjob_mkhomedir.so\fR module checks if the user's home directory
exists, and if it does not, it invokes the \fImkhomedirfor\fR
method of the \fI@NAMESPACE@.oddjob_mkhomedir\fR service for the \fBPAM_USER\fR
if the module is running with superuser privileges.  Otherwise, it invokes the
\fImkmyhomedir\fR method.

The location of the skeleton directory and the default umask are determined by
the configuration for the corresponding service in \fBoddjobd-mkhomedir.conf\fR,
so they can not be specified as arguments to this module.

If \fID-Bus\fR has not been configured to allow the calling application to
invoke these methods provided as part of the \fI@NAMESPACE@.oddjob_mkhomedir\fR
interface of the \fI/\fR object provided by the
\fI@NAMESPACE@.oddjob_mkhomedir\fR service, then \fIoddjobd\fR will not receive
the request and an error will be returned by \fID-Bus\fR.

If \fIoddjobd\fR has not been configured to provide these methods, or the
user is not authorized to use them, then \fIoddjobd\fR will ignore the
module's requests.

.SH SEE ALSO
\fBoddjob_request\fR(1)
\fBoddjob.conf\fR(5)
\fBoddjobd.conf\fR(5)
\fBoddjobd\fR(8)


src/NEWS
0.31.3 - Fix use of incorrectly-named facilities in the LSB header in the
         init script.
         Provide systemd configuration files.
         Some internal plumbing to possibly emit signals in addition to
	 reply messages in the future.
0.31.2 - Notice when we fail to register a name with the message bus.
0.31.1 - Tweak init script exit status when given an unrecognized command.
0.31   - Set permissions on intermediate directories correctly when the
         mkhomedir helper has to create some, using values from /etc/login.defs
         when no default umask is specified.
0.30.1 - Preprocess the man pages correctly so that mentions of the
         introspection interface and method names are correctly resolved to
	 their actual values.
0.30   - Drops the liboddjob wrapper library and python binding, which I don't
	 think anyone has been using.  (If you need either of these, I can add
	 them back.)


src/src/pam_oddjob_mkhomedir.c
/*
   Copyright 2005,2007,2008 Red Hat, Inc.
   All rights reserved.

   Redistribution and use in source and binary forms, with or without
   modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in
      the documentation and/or other materials provided with the
      distribution.
    * Neither the name of Red Hat, Inc., nor the names of its
      contributors may be used to endorse or promote products derived
      from this software without specific prior written permission.

   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
   IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
   TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
   PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER
   OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
   EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
   PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
   PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
   LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
   NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
   SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include "../config.h"
#include <sys/types.h>
#include <sys/stat.h>
#include <errno.h>
#include <pwd.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <security/pam_modules.h>
#include "common.h"
#include "oddjob_dbus.h"

static void
conv_text_info(pam_handle_t *pamh, const char *info)
{
	struct pam_conv *conv;
	int i;
	conv = NULL;
	if ((pam_get_item(pamh, PAM_CONV, (const void**) &conv) == PAM_SUCCESS) &&
	    (conv != NULL)) {
		const struct pam_message message = {
			.msg_style = PAM_TEXT_INFO,
			.msg = info,
		};
		const struct pam_message *messages[] = {
			&message,
			NULL,
		};
		struct pam_response *responses;
		if (conv->conv != NULL) {
			responses = NULL;
			i = conv->conv(1, messages,
				       &responses, conv->appdata_ptr);
			if ((i == PAM_SUCCESS) && (responses != NULL)) {
				if (responses->resp != NULL) {
					free(responses->resp);
				}
				free(responses);
			}
		}
	}
}

static void
send_pam_oddjob_mkhomedir_request(pam_handle_t *pamh)
{
	const char *user;
	char *buf, reply[8192];
	size_t bufsize;
	struct passwd pwd, *pw;
	struct stat st;
	int ret, result;

	user = NULL;
	strcpy(reply, "");

	if ((pam_get_user(pamh, &user, "login: ") == PAM_SUCCESS) &&
	    (user != NULL) &&
	    (strlen(user) > 0)) {
		/* Attempt to look up information about the user. */
		bufsize = 4;
		do {
			pw = NULL;
			buf = malloc(bufsize);
			if (buf == NULL) {
				break;
			}
			ret = getpwnam_r(user, &pwd, buf, bufsize, &pw);
			if ((ret != 0) || (pw != &pwd)) {
				pw = NULL;
				free(buf);
				buf = NULL;
				if (errno == ERANGE) {
					bufsize += 4;
				} else {
					break;
				}
			}
		} while ((ret != 0) && (errno == ERANGE));
		if ((pw != NULL) &&
		    (stat(pw->pw_dir, &st) == -1) && (errno == ENOENT)) {
			/* If we're running with the user's privileges, then
			 * call the "mkmyhomedir" method. */
			if ((getuid() == pw->pw_uid) &&
			    (geteuid() == pw->pw_uid) &&
			    (getgid() == pw->pw_gid) &&
			    (getegid() == pw->pw_gid)) {
				ret = oddjob_dbus_call_method(DBUS_BUS_SYSTEM,
							      ODDJOB_SERVICE_NAME "_mkhomedir",
							      "/",
							      ODDJOB_INTERFACE_NAME "_mkhomedir",
							      "mkmyhomedir",
							      &result,
							      reply,
							      sizeof(reply),
							      NULL,
							      0,
							      NULL);
			} else {
				/* Otherwise, attempt to have a directory
				 * created for the user by calling the
				 * "mkhomedirfor" method. */
				ret = oddjob_dbus_call_method(DBUS_BUS_SYSTEM,
							      ODDJOB_SERVICE_NAME "_mkhomedir",
							      "/",
							      ODDJOB_INTERFACE_NAME "_mkhomedir",
							      "mkhomedirfor",
							      &result,
							      reply,
							      sizeof(reply),
							      NULL,
							      0,
							      user,
							      NULL);
			}
		}
		if (buf != NULL) {
			free(buf);
		}
	}

	if (strlen(reply) > 0) {
		conv_text_info(pamh, reply);
	}
}

int
pam_sm_open_session(pam_handle_t *pamh, int flags, int argc, const char **argv)
{
	send_pam_oddjob_mkhomedir_request(pamh);
	return PAM_IGNORE;
}
int
pam_sm_close_session(pam_handle_t *pamh, int flags, int argc, const char **argv)
{
	return PAM_IGNORE;
}

int
pam_sm_acct_mgmt(pam_handle_t *pamh, int flags, int argc, const char **argv)
{
	send_pam_oddjob_mkhomedir_request(pamh);
	return PAM_IGNORE;
}
